File: tools/fleet_trend.py
#!/usr/bin/env python3
r"""fleet_trend — turn the fleet-status card from a one-off snapshot into a trend.

Every fleet-status card today answers only "what is true right now". The operator
cannot tell whether usable accounts are recovering or draining, whether the
escalation count is climbing, or whether live sessions just fell off a cliff — the
number carries no direction. This module is the missing time axis: each status tick
appends one compact row to a JSONL history ledger, and a pure renderer folds the
trailing rows into a sparkline + delta per load-bearing metric, so the card can show

    usable 3→3→2→1 ▇▇▄▁ (-2 vs 6 ticks)   escalate 0→0→1 ▁▁█ (+1)

instead of a bare ``usable 1``. It is deliberately tiny and dependency-free (stdlib
only, same as ``slack_post``): the ledger is append-only JSONL (the idiom the
loop-health fold already reads), bounded to a ring of the last ``cap`` rows so it can
never grow without limit, and the fold is pure given the rows + ``now`` so it is
table-tested with no clock and no disk.

  # append one tick from a fleet_top --json snapshot, then print the trend
  python tools/fleet_top.py --json | python tools/fleet_trend.py --append -
  python tools/fleet_trend.py --show          # render the trend from the ledger
  python tools/fleet_trend.py --show --json    # the machine-readable trend rows

The metrics extractor (``metrics_of``) reads a fleet_top snapshot, but ``append`` /
``tail`` / ``render`` take a plain flat ``{key: number}`` dict, so a second producer
(e.g. the dispatch card's queue depth / closed-per-hour) can feed the SAME ledger
later without touching this module.
"""
from __future__ import annotations

import argparse
import json
import os
import sys
from pathlib import Path
from typing import Any

SCHEMA = "fleet-trend/1"

# The default history ledger, workspace-relative. Runtime ticks belong under the
# gitignored state root so a scheduled status poll never dirties the shared tree.
DEFAULT_LEDGER = os.path.join(".fak", "nightrun", "fleet-status-history.jsonl")

# The load-bearing scalars a fleet-status trend is about, in render order: each is a
# (key, label) the renderer shows a sparkline + delta for. Chosen so the line reads as
# "is the fleet recovering or draining" at a glance, not a wall of every field.
METRICS: list[tuple[str, str]] = [
    ("usable", "usable"),      # worker accounts usable right now (capacity)
    ("live", "live"),          # live sessions (throughput in flight)
    # The window's session TOTAL (live + resumable + stuck + terminal-history). The
    # ledger key stays "sessions" for history compat, but the rendered label is the
    # explicit "all-sessions": a bare "sessions N→M" misreads as live fleet scale when
    # the count is terminal-history-dominated, so the trend never shows the unqualified
    # word next to the live series (#4651). "live" above is the actionable series.
    ("sessions", "all-sessions"),
    ("escalate", "escalate"),  # operator-actionable items the lifecycle can't heal
]

# Cumulative throughput counters, persisted alongside the gauges so the goodput /
# HEAD-stall SLO (computed by internal/fleettrend on read) can diff them across the
# window. They are NOT rendered by render_line — only recorded. A snapshot carrying no
# throughput block simply never sets them, so a counter-free history honestly reports
# its rates as absent rather than a fabricated zero. Key names mirror internal/fleettrend
# Counters exactly so the Go reader picks them straight off the ledger rows this writes.
COUNTERS: list[tuple[str, str]] = [
    ("lands", "lands"),                      # git-witnessed lands (trunk commit total)
    ("resumes", "resumes"),                  # auto-resume launches
    ("deaths", "deaths"),                    # worker crash/exhaust exits
    ("lands_witnessed", "lands_witnessed"),  # 1 when lands is git-witnessed, else absent
]

# Sparkline ramp, low→high. Eight levels is the most a proportional font renders
# distinctly on a phone; a flat series collapses to the lowest block.
_BLOCKS = "▁▂▃▄▅▆▇█"

# Default ring size — how many trailing ticks the ledger keeps. A daily/hourly tick
# stays well under this for months; the cap only guards against unbounded growth.
DEFAULT_CAP = 500


def metrics_of(snap: dict[str, Any]) -> dict[str, float]:
    """Extract the trend scalars from a fleet_top snapshot. Tolerant of a partial or
    error snapshot: a missing block reads 0, never raises."""
    sess = snap.get("sessions") or {}
    by_cat = sess.get("by_category") or {}
    acc = snap.get("accounts") or {}
    sysv = snap.get("system") or {}
    m = {
        "usable": float(acc.get("usable") or 0),
        "live": float(by_cat.get("LIVE") or 0),
        "sessions": float(sess.get("total") or 0),
        "escalate": float(sysv.get("escalate") or 0),
    }
    # Cumulative throughput counters ride in an optional "throughput" block; a snapshot
    # that carries none leaves them unset (honest absence). lands_witness is a provenance
    # string the producer stamps "git" when lands was counted from the commit graph — we
    # fold it to a 1 flag under the key internal/fleettrend reads.
    tp = snap.get("throughput")
    if isinstance(tp, dict) and tp:
        for c in ("lands", "resumes", "deaths"):
            if tp.get(c) is not None:
                m[c] = float(tp[c] or 0)
        if tp.get("lands_witness") in ("git", "witnessed"):
            m["lands_witnessed"] = 1.0
    return m


def append(path: str, metrics: dict[str, Any], now: str, *, cap: int = DEFAULT_CAP) -> dict[str, Any]:
    """Append one ``{ts, **metrics}`` row to the JSONL ledger and return it.

    Bounded: after appending, the file is trimmed to its last ``cap`` rows so the ledger
    is a ring, never unbounded. Best-effort and never raises on a read error (a corrupt
    prior line is dropped, not fatal) — a status tick must never crash on its own
    history. The directory is created if absent."""
    row = {"ts": now}
    # Backend-local dispatcher probes are not an aggregate fleet census. Preserve their
    # explicit scope for diagnostics without feeding backend_live into the aggregate
    # ``live`` trend or its net-decline alarm.
    if metrics.get("scope") == "backend":
        row["scope"] = "backend"
        row["backend"] = str(metrics.get("backend") or "unknown")
        row["backend_live"] = _num(metrics.get("backend_live"))
    for k, _ in METRICS:
        if k in metrics:
            row[k] = _num(metrics[k])
    # Persist the cumulative counters too when present, so the goodput / HEAD-stall SLO
    # is computable on read. Absent from `metrics` ⇒ absent from the row (never a 0).
    for k, _ in COUNTERS:
        if k in metrics:
            row[k] = _num(metrics[k])
    rows = _read_rows(path)
    rows.append(row)
    if cap > 0 and len(rows) > cap:
        rows = rows[-cap:]
    p = Path(path)
    if p.parent and not p.parent.exists():
        p.parent.mkdir(parents=True, exist_ok=True)
    with open(path, "w", encoding="utf-8") as fh:
        for r in rows:
            fh.write(json.dumps(r, separators=(",", ":")) + "\n")
    return row


def tail(path: str, n: int) -> list[dict[str, Any]]:
    """The last ``n`` ledger rows (oldest→newest). Missing/empty ledger → []."""
    rows = _read_rows(path)
    return rows[-n:] if n > 0 else rows


def spark(values: list[float]) -> str:
    """A unicode sparkline for a numeric series. A flat series (all equal, incl. a
    single point) renders as the lowest block per point — no false slope."""
    if not values:
        return ""
    lo, hi = min(values), max(values)
    if hi <= lo:
        return _BLOCKS[0] * len(values)
    span = hi - lo
    return "".join(_BLOCKS[min(len(_BLOCKS) - 1, int((v - lo) / span * (len(_BLOCKS) - 1) + 0.5))]
                   for v in values)


def _fmt(v: float) -> str:
    """Render a metric value: integers without a trailing ``.0``."""
    return str(int(v)) if float(v).is_integer() else f"{v:g}"


def metric_trend(rows: list[dict[str, Any]], key: str) -> dict[str, Any] | None:
    """Fold one metric across the rows into {first, last, delta, spark, n}. Returns
    None when the metric never appears (nothing to trend)."""
    series = [_num(r[key]) for r in rows if isinstance(r.get(key), (int, float))]
    if not series:
        return None
    return {
        "key": key,
        "first": series[0],
        "last": series[-1],
        "delta": round(series[-1] - series[0], 3),
        "spark": spark(series),
        "n": len(series),
    }


# Net-worker-decline alarm (#4591): how many CONSECUTIVE trailing appends of the
# `live` series must each strictly decline before the drain alarm fires. Three
# mirrors SPAWN_FAILED_RED_STREAK (tools/issue_resolve_dispatch.py): one dip is
# workers finishing, two can be a wind-down — three in a row with no recovery is
# the drain signature (a dead seat / dying pool bleeding the fleet toward zero).
NET_DECLINE_ALARM_STREAK = 3


def net_decline_alarm(rows: list[dict[str, Any]], *, key: str = "live",
                      streak: int = NET_DECLINE_ALARM_STREAK) -> dict[str, Any]:
    """Alarm when the trailing ``key`` series declines STRICTLY for ``streak``
    consecutive ledger appends (#4591) — the net-worker-decline half of the
    dead-seat drain fix: a fleet whose live count only ever steps down is
    draining, whatever each individual tick verdict said. Pure fold over the
    ledger rows (oldest→newest, the :func:`tail` shape). Returns
    ``{red, declines, key, first, last, n, reason}`` — ``declines`` is the
    length of the trailing strict-decline run, ``reason`` non-empty only when
    red. A short / missing / non-declining series is never red (fail-open)."""
    series = [_num(r[key]) for r in rows if isinstance(r.get(key), (int, float))]
    declines = 0
    for prev, cur in zip(reversed(series[:-1]), reversed(series[1:])):
        if cur < prev:
            declines += 1
        else:
            break
    red = declines >= max(1, int(streak))
    reason = ""
    if red:
        window = series[-(declines + 1):]
        reason = (f"net worker decline: '{key}' fell {declines} consecutive "
                  f"ledger appends ({'→'.join(_fmt(v) for v in window)}) — the "
                  f"fleet is draining, not cycling; check seat health "
                  f"(`fak accounts status`) and the dispatch card (#4591)")
    return {"red": red, "key": key, "declines": declines,
            "first": series[0] if series else None,
            "last": series[-1] if series else None,
            "n": len(series), "reason": reason}


def render_line(rows: list[dict[str, Any]]) -> str:
    """One compact trend line across all METRICS present in the rows: per metric a
    ``label first→last spark (Δ±d over N)``. Empty when there is no history yet, so the
    caller can drop the line entirely on a first-ever tick."""
    if not rows:
        return ""
    parts: list[str] = []
    for key, label in METRICS:
        t = metric_trend(rows, key)
        if t is None:
            continue
        arrow = f"{_fmt(t['first'])}→{_fmt(t['last'])}" if t["n"] > 1 else _fmt(t["last"])
        delta = ""
        if t["n"] > 1 and t["delta"]:
            sign = "+" if t["delta"] > 0 else ""
            delta = f" ({sign}{_fmt(t['delta'])} over {t['n']})"
        parts.append(f"{label} {arrow} {t['spark']}{delta}".rstrip())
    return "trend: " + " · ".join(parts) if parts else ""


def _num(v: Any) -> float:
    try:
        f = float(v)
    except (TypeError, ValueError):
        return 0.0
    return int(f) if f.is_integer() else f


def _read_rows(path: str) -> list[dict[str, Any]]:
    try:
        with open(path, encoding="utf-8") as fh:
            text = fh.read()
    except OSError:
        return []
    out: list[dict[str, Any]] = []
    for line in text.split("\n"):
        line = line.strip()
        if not line:
            continue
        try:
            row = json.loads(line)
        except ValueError:
            continue  # tolerate a torn line rather than brick the history
        if isinstance(row, dict):
            out.append(row)
    return out


def _iso_now() -> str:
    import datetime as dt
    return dt.datetime.now(dt.timezone.utc).isoformat(timespec="seconds").replace("+00:00", "Z")


def main(argv: list[str] | None = None) -> int:
    ap = argparse.ArgumentParser(description="Append + render the fleet-status trend ledger.")
    ap.add_argument("--ledger", default=DEFAULT_LEDGER, help="JSONL history path")
    ap.add_argument("--append", metavar="SNAPSHOT", default="",
                    help="append a tick from a fleet_top --json snapshot (file or - for stdin)")
    ap.add_argument("--show", action="store_true", help="render the trend from the ledger")
    ap.add_argument("--window", type=int, default=24, help="how many trailing ticks to fold")
    ap.add_argument("--cap", type=int, default=DEFAULT_CAP, help="max ledger rows (ring size)")
    ap.add_argument("--now", default="", help="override the tick timestamp (tests)")
    ap.add_argument("--json", action="store_true", help="emit machine-readable rows")
    args = ap.parse_args(argv)

    try:
        sys.stdout.reconfigure(encoding="utf-8")  # type: ignore[union-attr]
    except (AttributeError, ValueError):
        pass

    if args.append:
        raw = sys.stdin.read() if args.append == "-" else Path(args.append).read_text(encoding="utf-8")
        snap = json.loads(raw)
        row = append(args.ledger, metrics_of(snap), args.now or _iso_now(), cap=args.cap)
        print(json.dumps(row) if args.json else f"appended {row}")
        if not args.show:
            return 0

    rows = tail(args.ledger, args.window)
    if args.json:
        print(json.dumps({"schema": SCHEMA, "ledger": args.ledger, "rows": rows}, indent=2))
    else:
        line = render_line(rows)
        print(line or "(no history yet)")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: internal/fleettrend/fleettrend.go
package fleettrend

import (
	"bufio"
	"bytes"
	"encoding/json"
	"fmt"
	"math"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/anthony-chaudhary/fak/internal/jsonlledger"
)

// Schema is the versioned schema identifier for fleet-trend records.
const Schema = "fleet-trend/1"

// DefaultLedger is the default repository-relative path to the fleet status history JSONL ledger.
const DefaultLedger = ".fak/nightrun/fleet-status-history.jsonl"

// DefaultCap is the default maximum number of historical rows retained in the trend ledger.
const DefaultCap = 500

// MetricDef pairs a metric key with its display label.
type MetricDef struct {
	Key   string
	Label string
}

// Metrics defines the default gauge metrics tracked and rendered in trend lines.
var Metrics = []MetricDef{
	{Key: "usable", Label: "usable"},
	{Key: "live", Label: "live"},
	{Key: "sessions", Label: "sessions"},
	{Key: "escalate", Label: "escalate"},
}

// Counters are monotonic cumulative totals persisted alongside the gauges so the
// throughput SLO (WindowRates) can diff them across a rolling window. They are
// NOT rendered by RenderLine — only the rate derivation reads them. A snapshot
// that omits the throughput object simply never sets them, so a window without
// counters honestly reports its rates as absent rather than a fabricated zero.
var Counters = []MetricDef{
	{Key: "lands", Label: "lands"},     // git-WITNESSED lands (see landsWitnessedKey)
	{Key: "resumes", Label: "resumes"}, // auto-resume launches
	{Key: "deaths", Label: "deaths"},   // worker crash/exhaust exits
	{Key: landsWitnessedKey, Label: "lands_witnessed"},
}

// landsWitnessedKey persists whether the lands counter is git-witnessed (1) or
// self-reported (0/absent). Goodput leans on lands, so the render surfaces this
// provenance rather than letting a self-reported number pass as witnessed. Ties
// to the OBSERVED-vs-WITNESSED provenance tag the upstream snapshot must fill.
const landsWitnessedKey = "lands_witnessed"

var blocks = []rune("▁▂▃▄▅▆▇█")

// MetricsOf extracts gauge metrics and optional throughput counters from a raw fleet status snapshot.
func MetricsOf(snap map[string]any) map[string]float64 {
	sessions := asMap(snap["sessions"])
	byCategory := asMap(sessions["by_category"])
	accounts := asMap(snap["accounts"])
	system := asMap(snap["system"])
	m := map[string]float64{
		"usable":   number(accounts["usable"]),
		"live":     number(byCategory["LIVE"]),
		"sessions": number(sessions["total"]),
		"escalate": number(system["escalate"]),
	}
	// Cumulative throughput counters live under an optional "throughput" object so
	// a health snapshot that carries none leaves the rate columns unset (honest
	// absence). lands_witness is a provenance string the producer stamps "git"
	// when the lands total was counted from origin/main-ancestor commits.
	if tp := asMap(snap["throughput"]); len(tp) > 0 {
		for _, c := range []string{"lands", "resumes", "deaths"} {
			if v, ok := tp[c]; ok {
				m[c] = number(v)
			}
		}
		if w, _ := tp["lands_witness"].(string); w == "git" || w == "witnessed" {
			m[landsWitnessedKey] = 1
		}
	}
	return m
}

// Append records a new timestamped metric row to the ledger, capping total rows to capRows.
func Append(path string, metrics map[string]float64, now string, capRows int) (map[string]any, error) {
	row := map[string]any{"ts": now}
	for _, metric := range Metrics {
		if v, ok := metrics[metric.Key]; ok {
			row[metric.Key] = compactNumber(v)
		}
	}
	for _, counter := range Counters {
		if v, ok := metrics[counter.Key]; ok {
			row[counter.Key] = compactNumber(v)
		}
	}
	rows := readRows(path)
	rows = append(rows, row)
	if capRows > 0 && len(rows) > capRows {
		rows = rows[len(rows)-capRows:]
	}
	if dir := filepath.Dir(path); dir != "." && dir != "" {
		if err := os.MkdirAll(dir, 0o755); err != nil {
			return nil, err
		}
	}
	f, err := os.Create(path)
	if err != nil {
		return nil, err
	}
	defer f.Close()
	enc := json.NewEncoder(f)
	enc.SetEscapeHTML(false)
	for _, r := range rows {
		if err := enc.Encode(r); err != nil {
			return nil, err
		}
	}
	return row, nil
}

// Tail returns the trailing n rows from the ledger, or all rows if n <= 0.
func Tail(path string, n int) []map[string]any {
	rows := foldedRows(path)
	if n <= 0 || n >= len(rows) {
		return rows
	}
	return rows[len(rows)-n:]
}

// Spark renders a compact unicode sparkline string representing the trajectory of a numeric slice.
func Spark(values []float64) string {
	if len(values) == 0 {
		return ""
	}
	lo, hi := values[0], values[0]
	for _, v := range values[1:] {
		if v < lo {
			lo = v
		}
		if v > hi {
			hi = v
		}
	}
	if hi <= lo {
		return strings.Repeat(string(blocks[0]), len(values))
	}
	span := hi - lo
	var b strings.Builder
	for _, v := range values {
		idx := int((v-lo)/span*float64(len(blocks)-1) + 0.5)
		if idx < 0 {
			idx = 0
		}
		if idx >= len(blocks) {
			idx = len(blocks) - 1
		}
		b.WriteRune(blocks[idx])
	}
	return b.String()
}

// Trend summarizes the trajectory of a single metric across a series of ledger rows.
type Trend struct {
	Key   string  `json:"key"`
	First float64 `json:"first"`
	Last  float64 `json:"last"`
	Delta float64 `json:"delta"`
	Spark string  `json:"spark"`
	N     int     `json:"n"`
}

// MetricTrend calculates the start value, end value, delta, and sparkline for a given metric key across rows.
func MetricTrend(rows []map[string]any, key string) (Trend, bool) {
	series := make([]float64, 0, len(rows))
	for _, row := range rows {
		switch v := row[key].(type) {
		case int:
			series = append(series, float64(v))
		case int64:
			series = append(series, float64(v))
		case float64:
			series = append(series, v)
		case json.Number:
			if f, err := v.Float64(); err == nil {
				series = append(series, f)
			}
		}
	}
	if len(series) == 0 {
		return Trend{}, false
	}
	delta := math.Round((series[len(series)-1]-series[0])*1000) / 1000
	return Trend{
		Key:   key,
		First: series[0],
		Last:  series[len(series)-1],
		Delta: delta,
		Spark: Spark(series),
		N:     len(series),
	}, true
}

// RenderLine formats a one-line summary of gauge metric trends across the provided rows.
func RenderLine(rows []map[string]any) string {
	if len(rows) == 0 {
		return ""
	}
	var parts []string
	for _, metric := range Metrics {
		trend, ok := MetricTrend(rows, metric.Key)
		if !ok {
			continue
		}
		arrow := formatNumber(trend.Last)
		if trend.N > 1 {
			arrow = formatNumber(trend.First) + "→" + formatNumber(trend.Last)
		}
		delta := ""
		if trend.N > 1 && trend.Delta != 0 {
			sign := ""
			if trend.Delta > 0 {
				sign = "+"
			}
			delta = fmt.Sprintf(" (%s%s over %d)", sign, formatNumber(trend.Delta), trend.N)
		}
		parts = append(parts, strings.TrimSpace(fmt.Sprintf("%s %s %s%s", metric.Label, arrow, trend.Spark, delta)))
	}
	if len(parts) == 0 {
		return ""
	}
	return "trend: " + strings.Join(parts, " · ")
}

// Rate is one per-hour figure derived over the rolling window. Present is false
// when the window lacks the two counter-bearing, timestamped endpoints needed to
// derive it, so callers render "n/a" instead of a fabricated 0/hr.
type Rate struct {
	PerHour float64 `json:"per_hour"`
	Delta   float64 `json:"delta"`
	Present bool    `json:"present"`
}

// Rates is the fleet-throughput SLO over a rolling window: how fast work lands
// against how fast workers resume and die. Goodput = landsΔ / (landsΔ + deathsΔ)
// answers whether adding workers shipped work or just churned. LandsWitnessed
// carries whether the lands total is git-witnessed (honest) or self-reported.
type Rates struct {
	WindowHours    float64 `json:"window_hours"`
	Ticks          int     `json:"ticks"`
	Lands          Rate    `json:"lands"`
	Resumes        Rate    `json:"resumes"`
	Deaths         Rate    `json:"deaths"`
	Goodput        float64 `json:"goodput"`
	GoodputPresent bool    `json:"goodput_present"`
	LandsWitnessed bool    `json:"lands_witnessed"`
}

// WindowRates derives resumes/hr, deaths/hr, lands/hr, and a goodput ratio from
// the cumulative counters in the given rows (already trimmed to the window by
// Tail). Each rate diffs the first and last row that carries its counter and
// divides by the wall-clock hours between their timestamps — so the arithmetic
// is a pure function of the ledger, no clock read. The second return is false
// when no counter could be derived (a window with no throughput data at all).
func WindowRates(rows []map[string]any) (Rates, bool) {
	out := Rates{Ticks: len(rows)}
	if len(rows) > 0 {
		if first, ok1 := parseRowTime(rows[0]["ts"]); ok1 {
			if last, ok2 := parseRowTime(rows[len(rows)-1]["ts"]); ok2 {
				if h := last.Sub(first).Hours(); h > 0 {
					out.WindowHours = h
				}
			}
		}
	}
	out.Lands = counterRate(rows, "lands")
	out.Resumes = counterRate(rows, "resumes")
	out.Deaths = counterRate(rows, "deaths")
	if out.Lands.Present && out.Deaths.Present {
		switch denom := out.Lands.Delta + out.Deaths.Delta; {
		case denom > 0:
			// Some work resolved this window (shipped and/or died): the real ship-vs-churn ratio.
			out.Goodput = out.Lands.Delta / denom
			out.GoodputPresent = true
		case out.Resumes.Present && out.Resumes.Delta > 0:
			// Nothing shipped AND nothing died, yet workers were actively resuming this window:
			// that is a real 0% conversion (the fleet spun without shipping), NOT "no data". Only
			// n/a when there is truly no activity to convert — an idle window with no resumes.
			out.Goodput = 0
			out.GoodputPresent = true
		}
	}
	out.LandsWitnessed = lastCounterFlag(rows, landsWitnessedKey)
	return out, out.Lands.Present || out.Resumes.Present || out.Deaths.Present
}

// counterRate reads the first and last timestamped rows that carry key and
// returns their delta over the hours between them. It needs two such rows: a
// single data point yields no rate. A negative delta (a counter reset when the
// producing process restarted) clamps to 0 rather than emitting a spurious rate.
func counterRate(rows []map[string]any, key string) Rate {
	var first, last float64
	var firstTime, lastTime time.Time
	var n int
	for _, row := range rows {
		v, ok := counterValue(row[key])
		if !ok {
			continue
		}
		ts, tsok := parseRowTime(row["ts"])
		if !tsok {
			continue
		}
		if n == 0 {
			first, firstTime = v, ts
		}
		last, lastTime = v, ts
		n++
	}
	if n < 2 {
		return Rate{}
	}
	hours := lastTime.Sub(firstTime).Hours()
	if hours <= 0 {
		return Rate{}
	}
	delta := last - first
	if delta < 0 {
		delta = 0
	}
	return Rate{PerHour: delta / hours, Delta: delta, Present: true}
}

// lastCounterFlag reports whether the most recent row carrying key sets it to a
// truthy (>= 1) value — used to read the lands-provenance flag off the freshest
// snapshot in the window.
func lastCounterFlag(rows []map[string]any, key string) bool {
	for i := len(rows) - 1; i >= 0; i-- {
		if v, ok := counterValue(rows[i][key]); ok {
			return v >= 1
		}
	}
	return false
}

// RenderThroughput is the one fleet-status row: the four SLO figures plus the
// window label. Absent counters render "n/a" (not 0/hr), and the lands figure
// carries a provenance tag so a self-reported goodput never reads as witnessed.
func RenderThroughput(rows []map[string]any) string {
	r, ok := WindowRates(rows)
	if !ok {
		return ""
	}
	goodput := "n/a"
	if r.GoodputPresent {
		goodput = fmt.Sprintf("%.0f%%", r.Goodput*100)
	}
	line := fmt.Sprintf("throughput: lands %s · resumes %s · deaths %s · goodput %s (%s)",
		fmtRate(r.Lands), fmtRate(r.Resumes), fmtRate(r.Deaths), goodput, windowLabel(r))
	if r.Lands.Present {
		if r.LandsWitnessed {
			line += " [lands: git-witnessed]"
		} else {
			line += " [lands: self-reported]"
		}
	}
	if stall := HeadStallTicks(rows); stall > 0 {
		// The fleet is spinning without advancing HEAD — surfaced only when it is actually
		// happening, so a healthy window's line is unchanged.
		line += fmt.Sprintf(" [HEAD stalled %d %s]", stall, pluralTicks(stall))
	}
	return line
}

// HeadStallTicks reports the number of trailing consecutive ticks that landed NO new work while
// workers were still active (resuming) — the fleet is spinning without advancing HEAD (Obs #2). It
// reads the per-tick lands counter (git-witnessed commits when LandsWitnessed is set, so a flat run
// is a literal HEAD stall) alongside resumes: the maximal trailing run where lands stays flat but
// resumes climbs. It returns 0 when HEAD advanced on the latest tick, when no worker was active to
// stall on, or when fewer than two lands-bearing snapshots exist — so a quiet fleet never reads as
// stalled. Pure function of the ledger rows; no clock or git read.
func HeadStallTicks(rows []map[string]any) int {
	var lands, resumes []float64
	for _, row := range rows {
		l, ok := counterValue(row["lands"])
		if !ok {
			continue
		}
		r, _ := counterValue(row["resumes"]) // absent → 0; a stall needs a positive resumes delta below
		lands = append(lands, l)
		resumes = append(resumes, r)
	}
	if len(lands) < 2 {
		return 0
	}
	stall := 0
	for i := len(lands) - 1; i > 0; i-- {
		// HEAD flat (no land) AND a worker resumed since the prior snapshot = one stalled tick.
		if lands[i] == lands[i-1] && resumes[i] > resumes[i-1] {
			stall++
			continue
		}
		break
	}
	return stall
}

func pluralTicks(n int) string {
	if n == 1 {
		return "tick"
	}
	return "ticks"
}

func fmtRate(r Rate) string {
	if !r.Present {
		return "n/a"
	}
	return fmt.Sprintf("%.1f/hr", r.PerHour)
}

func windowLabel(r Rates) string {
	ticks := fmt.Sprintf("%d ticks", r.Ticks)
	if r.Ticks == 1 {
		ticks = "1 tick"
	}
	if r.WindowHours <= 0 {
		return "over " + ticks
	}
	return fmt.Sprintf("over %.1fh · %s", r.WindowHours, ticks)
}

// parseRowTime parses a row's RFC3339 ts stamp (the format ISONow writes). A
// non-timestamp value (older torn rows, test fixtures) yields ok=false so the
// rate derivation skips it rather than crashing.
func parseRowTime(v any) (time.Time, bool) {
	s, ok := v.(string)
	if !ok || s == "" {
		return time.Time{}, false
	}
	t, err := time.Parse(time.RFC3339, s)
	if err != nil {
		return time.Time{}, false
	}
	return t, true
}

// counterValue reads a numeric counter, reporting presence separately from value
// so a missing key is distinguishable from a real 0 (number() folds both to 0).
func counterValue(v any) (float64, bool) {
	switch x := v.(type) {
	case int:
		return float64(x), true
	case int64:
		return float64(x), true
	case float64:
		return x, true
	case json.Number:
		f, err := x.Float64()
		return f, err == nil
	}
	return 0, false
}

// ISONow returns the current UTC time formatted as an RFC3339 string truncated to the second.
func ISONow() string {
	return time.Now().UTC().Truncate(time.Second).Format(time.RFC3339)
}

// readRows reads the whole ledger from scratch. The write path (Append) uses it
// so the byte-for-byte rewrite it performs never depends on cached fold state.
func readRows(path string) []map[string]any {
	f, err := os.Open(path)
	if err != nil {
		return nil
	}
	defer f.Close()
	scanner := bufio.NewScanner(f)
	var rows []map[string]any
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" {
			continue
		}
		dec := json.NewDecoder(strings.NewReader(line))
		dec.UseNumber()
		var row map[string]any
		if err := dec.Decode(&row); err != nil {
			continue
		}
		rows = append(rows, row)
	}
	return rows
}

// foldState accumulates the parsed rows in file order — the same slice readRows
// returns, so Tail/RenderLine see identical data.
type foldState struct {
	rows []map[string]any
}

var (
	foldMu    sync.Mutex
	foldCkpts = map[string]jsonlledger.Checkpoint[foldState]{}
)

// foldRow decodes one JSONL line exactly as readRows did — UseNumber so integer
// counts round-trip as json.Number, malformed lines dropped — and folds it in.
func foldRow(s foldState, raw json.RawMessage) foldState {
	dec := json.NewDecoder(bytes.NewReader(raw))
	dec.UseNumber()
	var row map[string]any
	if err := dec.Decode(&row); err != nil {
		return s
	}
	s.rows = append(s.rows, row)
	return s
}

// foldedRows is the read/render path readRows used to serve: it returns every
// ledger row, but folds only the bytes appended since the last read of this
// path via jsonlledger.TailFold, keyed by an in-process checkpoint. The dozens
// of panes that poll the trend re-read only the delta instead of the whole
// capped ledger each tick. A rewrite (the ledger dropping its oldest row at cap)
// changes the bytes ending at the prior offset, so TailFold re-folds in full —
// the output stays byte-identical to a from-scratch read. On any I/O error it
// falls back to that from-scratch read, matching readRows' best-effort contract.
func foldedRows(path string) []map[string]any {
	key := path
	if abs, err := filepath.Abs(path); err == nil {
		key = abs
	}
	foldMu.Lock()
	defer foldMu.Unlock()
	ck, err := jsonlledger.TailFold(path, foldCkpts[key], foldState{}, foldRow)
	if err != nil {
		delete(foldCkpts, key)
		return readRows(path)
	}
	foldCkpts[key] = ck
	if len(ck.State.rows) == 0 {
		return nil
	}
	// Hand back a copy: callers (and Append) mutate the slice they receive, and
	// the cached fold state must not move under a later delta read.
	out := make([]map[string]any, len(ck.State.rows))
	copy(out, ck.State.rows)
	return out
}

// resetFoldCache clears the in-process checkpoint cache. Tests that reuse a
// ledger path across scenarios call it to fold from a clean slate.
func resetFoldCache() {
	foldMu.Lock()
	defer foldMu.Unlock()
	foldCkpts = map[string]jsonlledger.Checkpoint[foldState]{}
}

func asMap(v any) map[string]any {
	if m, ok := v.(map[string]any); ok {
		return m
	}
	return map[string]any{}
}

func number(v any) float64 {
	switch x := v.(type) {
	case int:
		return float64(x)
	case int64:
		return float64(x)
	case float64:
		return x
	case json.Number:
		f, _ := x.Float64()
		return f
	case string:
		f, _ := strconv.ParseFloat(x, 64)
		return f
	default:
		return 0
	}
}

func compactNumber(v float64) any {
	if math.IsNaN(v) || math.IsInf(v, 0) {
		return 0
	}
	if math.Trunc(v) == v {
		return int(v)
	}
	return v
}

func formatNumber(v float64) string {
	if math.Trunc(v) == v {
		return strconv.FormatInt(int64(v), 10)
	}
	return strconv.FormatFloat(v, 'g', -1, 64)
}
